core/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/entity/decorator/FilteringEntityByEntitySelector.java

```diff
@@ -0,0 +1,411 @@
+                return false;
+            }
+            var replayedValue = variableDescriptor.getValue(replayedEntity);
+            return replayedValue == null || valueRangeManager
+                    .getFromEntity(variableDescriptor.getValueRangeDescriptor(), otherEntity).contains(replayedValue);
+        }
+    }
+
+    private static class OriginalFilteringValueRangeIterator<Solution_> extends AbstractFilteringValueRangeIterator<Solution_>
+            implements ListIterator<Object> {
+
+        private final ListIterator<Object> entityIterator;
+        private Object selected = null;
+
+        private OriginalFilteringValueRangeIterator(Supplier<Object> upcomingEntitySupplier,
+                ListIterator<Object> entityIterator, BasicVariableDescriptor<Solution_>[] basicVariableDescriptors,
+                ValueRangeManager<Solution_> valueRangeManager) {
+            super(upcomingEntitySupplier, basicVariableDescriptors, valueRangeManager);
+            this.entityIterator = entityIterator;
+        }
+
+        @Override
+        public boolean hasNext() {
+            this.selected = pickNext();
+            return selected != null;
+        }
+
+        private Object pickNext() {
+            if (selected != null) {
+                throw new IllegalStateException("The next value has already been picked.");
+            }
+            initialize();
+            this.selected = null;
+            while (entityIterator.hasNext()) {
+                var entity = entityIterator.next();
+                if (isReachable(entity)) {
+                    return entity;
+                }
+            }
+            return null;
+        }
+
+        private Object pickSelected() {
+            if (selected == null) {
+                throw new NoSuchElementException();
+            }
+            var result = selected;
+            this.selected = null;
+            return result;
+        }
+
+        @Override
+        public Object next() {
+            return pickSelected();
+        }
+
+        @Override
+        public boolean hasPrevious() {
+            this.selected = pickPrevious();
+            return selected != null;
+        }
+
+        private Object pickPrevious() {
+            if (selected != null) {
+                throw new IllegalStateException("The next value has already been picked.");
+            }
+            initialize();
+            this.selected = null;
+            while (entityIterator.hasPrevious()) {
+                var entity = entityIterator.previous();
+                if (isReachable(entity)) {
+                    return entity;
+                }
+            }
+            return null;
+        }
+
+        @Override
+        public Object previous() {
+            return pickSelected();
+        }
+
+        @Override
+        public int nextIndex() {
+            return entityIterator.nextIndex();
+        }
+
+        @Override
+        public int previousIndex() {
```

Comment:
Do we have any actual use for previous(), previousIndex() etc.? Where is it used?
IMO this should not be implemented, we will never use these iterators to go back in time.

Arguably, all methods of these iterators which we do not need should just throw `UnsupportedOperationException`. The selectors are complex enough, even without implementing methods we do not need.